File: 105. Construct Binary Tree from Preorder and Inorder Traversal/C++/Solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode *buildTree(vector<int> &preorder, vector<int> &inorder) {
        if (preorder.size() == 0 || inorder.size() == 0) {
            return nullptr;
        }
        TreeNode *node = new TreeNode(preorder.front());
        preorder.erase(preorder.begin());
        int index;
        for (int i = 0; i < inorder.size(); i++) {
            if (inorder[i] == node->val) {
                index = i;
                break;
            }
        }
        vector<int> left(inorder.begin(), inorder.begin() + index);
        node->left = buildTree(preorder, left);
        vector<int> right(inorder.begin() + index + 1, inorder.end());
        node->right = buildTree(preorder, right);
        return node;
    }
};

File: 105. Construct Binary Tree from Preorder and Inorder Traversal/Python/Solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution(object):
    def buildTree(self, preorder, inorder):
        """
        :type preorder: List[int]
        :type inorder: List[int]
        :rtype: TreeNode
        """
        if not preorder or not inorder:
            return None
        node = TreeNode(preorder.pop(0))
        index = inorder.index(node.val)
        node.left = self.buildTree(preorder, inorder[:index])
        node.right = self.buildTree(preorder, inorder[index + 1:])
        return node
